### PR TITLE
Added a setting to change whether a not logged in reminder shoul…

### PR DESCRIPTION
**Pull request by frederikmillingpytlick**

### CODE REVIEW

File: db/install.xml [hunk 1]
```diff
@@ -25,6 +25,22 @@
...
+        <FIELD NAME="id" TYPE="int" LENGTH="18" NOTNULL="true" SEQUENCE="true"/>
+        <FIELD NAME="userid" TYPE="int" LENGTH="18" NOTNULL="true" SEQUENCE="false"/>
+        <FIELD NAME="reminderid" TYPE="int" LENGTH="18" NOTNULL="true" SEQUENCE="false"/>
+        <FIELD NAME="timesent" TYPE="int" LENGTH="18" NOTNULL="true" SEQUENCE="false"/>
+      </FIELDS>
+      <KEYS>
+        <KEY NAME="id" TYPE="primary" FIELDS="id"/>
+      </KEYS>
+      <INDEXES>
+        <INDEX NAME="userid_reminderid_ix" UNIQUE="false" FIELDS="userid,reminderid"/>
```
Comment (johnnypraxis) [inline db/install.xml:39]: Shouldn't this index be unique? Can a user have multiple reminders?

File: db/upgrade.php [hunk 2]
```diff
@@ -189,5 +189,25 @@ function xmldb_local_augmented_teacher_upgrade($oldversion) {
...
+    if ($oldversion < 2022073000) {
+        $table = new xmldb_table('local_augmented_teacher_sent');
+
+        $table->add_field('id', XMLDB_TYPE_INTEGER, '18', null, XMLDB_NOTNULL, XMLDB_SEQUENCE, null);
+        $table->add_field('userid', XMLDB_TYPE_INTEGER, '18', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('reminderid', XMLDB_TYPE_INTEGER, '18', null, XMLDB_NOTNULL, null, null);
+        $table->add_field('timesent', XMLDB_TYPE_INTEGER, '18', null, XMLDB_NOTNULL, null, '0');
+
+        $table->add_key('id', XMLDB_KEY_PRIMARY, ['id']);
+        $table->add_index('userid_reminderid_ix', XMLDB_INDEX_NOTUNIQUE, ['userid', 'reminderid']);
```
Comment (johnnypraxis) [inline db/upgrade.php:202]: Same as previous comment.

File: lib.php [hunk 3]
```diff
@@ -347,10 +355,11 @@ function local_augmented_teacher_send_notloggedin_reminder_message() {
                         $log->timecreated = time();
                         $DB->insert_record('local_augmented_teacher_log', $log);
 
-                        // Set sent status
-                        if ($has_sent === false) {
-                            $has_sent = true;
-                        }
+                        $DB->insert_record('local_augmented_teacher_sent', (object)[
```
Comment (johnnypraxis) [inline lib.php:358]: The record should be inserted with an entity in a repository.

Example:
$repo = $base_factory->etc()->etc()->get_repository();
$repo->insert($entity);

File: lib.php [hunk 4]
```diff
@@ -364,11 +373,18 @@ function local_augmented_teacher_send_notloggedin_reminder_message() {
...
-    if ($has_sent) {
-        set_config('notloggedinlast', $timenow, 'local_augmented_teacher');
-    }
+function get_last_time_reminder_was_sent_to_user(int $reminderid, int $userid): int {
+    global $DB;
+
+    $sent_reminders = $DB->get_records('local_augmented_teacher_sent', ['userid' => $userid, 'reminderid' => $reminderid], 'id DESC');
+
+    return array_values($sent_reminders)[0]->timesent ?? 0;
+}
```
Comment (johnnypraxis) [inline lib.php:384]: To repo.

**Review by johnnypraxis — COMMENTED** [review]
See questions. Needs a bit of refactoring and tests as well.

File: db/install.xml (hunk 1 above)
Comment (frederikmillingpytlick) [reply db/install.xml:39]: A user can receive the same reminder multiple times so it should not be unique

File: db/upgrade.php (hunk 2 above)
Comment (frederikmillingpytlick) [reply db/upgrade.php:202]: A user can receive the same reminder multiple times so it should not be unique

**Review by johnnypraxis — APPROVED** [review]

**Comment by superfoght** [comment]
Test and confirmed it works 🚀

**Comment by superfoght** [comment]
Tested and confirmed it works 🚀

**Pull request merged**